src/meta.rs

```diff
@@ -2,21 +2,78 @@ use crate::id::ResourceId;
 
 use anyhow::Error;
 use canonical_path::CanonicalPathBuf;
+use serde::{Deserialize, Serialize};
+use std::any;
 use std::ffi::{OsStr, OsString};
+use std::fs::{self, File};
+use std::io::Write;
+use std::path::Path;
 use std::time::SystemTime;
 use walkdir::DirEntry;
 
-#[derive(Eq, PartialEq, Hash, Clone, Debug)]
+const METADATA_RELATIVE_PATH: &str = ".ark/meta";
+
+#[derive(Eq, PartialEq, Hash, Clone, Debug, Deserialize, Serialize)]
 pub struct ResourceMeta {
     pub id: ResourceId,
-    pub modified: SystemTime,
+    // pub modified: SystemTime,
+    pub created_date: SystemTime,
     pub name: Option<OsString>,
     pub extension: Option<OsString>,
     pub kind: Option<ResourceKind>,
     pub extra: Option<ResourceExtra>,
 }
 
 impl ResourceMeta {
+    pub fn store<P: AsRef<Path>>(
+        root: P,
+        path: P,
+        resource_id: ResourceId,
+        name: Option<OsString>,
+        extension: Option<OsString>,
+        kind: Option<ResourceKind>,
+        extra: Option<ResourceExtra>,
+    ) {
+        let metadata_path = root.as_ref().join(METADATA_RELATIVE_PATH);
+        fs::create_dir_all(metadata_path.to_owned())
+            .expect(&format!("Creating {} directory", METADATA_RELATIVE_PATH));
+        let mut metadata_file = File::create(
+            metadata_path
+                .to_owned()
+                .join(format!("{}", resource_id.crc32)),
```

Comment:
Whole `resource_id` must be used in the filename here.
E.g. `123-45678` where 123 is size of the data, and 45678 is CRC-32 checksum.